cardano-cli/src/Cardano/CLI/Types/Errors/TxValidationError.hs

```diff
@@ -70,64 +65,80 @@ validateScriptSupportedInEra era script@(ScriptInAnyLang lang _) =
+data TxNotSupportedInEraValidationError era =
+    -- | First argument is the kind of data that is not supported.
+    -- Second argument in the era that doesn't support the data.
+    TxNotSupportedInAnyCardanoEraValidationError T.Text AnyCardanoEra
+    -- | First argument is the kind of data that is not supported.
+    -- Second argument in the era that doesn't support the data.
+  | TxNotSupportedInShelleyBasedEraValidationError T.Text (ShelleyBasedEra era)
+
+instance Show (TxNotSupportedInEraValidationError era) where
+  show =
+    \case
+      TxNotSupportedInAnyCardanoEraValidationError a cEra -> go a cEra
+      TxNotSupportedInShelleyBasedEraValidationError a sbe -> go a sbe
+    where
+      go a era = show (pretty a) <> " not supported in " <> show era
+
+instance Error (TxNotSupportedInEraValidationError era) where
+  prettyError =
+    \case
+      TxNotSupportedInAnyCardanoEraValidationError a cEra -> go a cEra
+      TxNotSupportedInShelleyBasedEraValidationError a sbe -> go a sbe
+    where
+      go a cEra = pretty a <+> "not supported in" <+> viaShow cEra
+
+validateTxTotalCollateral :: ShelleyBasedEra era
                           -> Maybe L.Coin
-                          -> Either TxTotalCollateralValidationError (TxTotalCollateral era)
+                          -> Either (TxNotSupportedInEraValidationError era) (TxTotalCollateral era)
 validateTxTotalCollateral _ Nothing = return TxTotalCollateralNone
-validateTxTotalCollateral era (Just coll) = do
-  supported <- conjureWitness era TxTotalCollateralNotSupported
+validateTxTotalCollateral sbe (Just coll) = do
+  supported <- conjureWitness (toCardanoEra sbe) $ TxNotSupportedInAnyCardanoEraValidationError "Transaction collateral"
   pure $ TxTotalCollateral supported coll
 
-newtype TxReturnCollateralValidationError
-  = TxReturnCollateralNotSupported AnyCardanoEra
-  deriving Show
-
-instance Error TxReturnCollateralValidationError where
-  prettyError (TxReturnCollateralNotSupported era) =
-    "Transaction return collateral not supported in " <> pretty era
-
-validateTxReturnCollateral :: CardanoEra era
+validateTxCurrentTreasuryValue :: ()
+  => ShelleyBasedEra era
+  -> Maybe TxCurrentTreasuryValue
+  -> Either (TxNotSupportedInEraValidationError era) (Maybe (Featured ConwayEraOnwards era L.Coin))
+validateTxCurrentTreasuryValue sbe mCurrentTreasuryValue =
+  case mCurrentTreasuryValue of
+    Nothing -> Right Nothing
+    Just (TxCurrentTreasuryValue { unTxCurrentTreasuryValue }) ->
+      caseShelleyToBabbageOrConwayEraOnwards
+        (const $ Left $ TxNotSupportedInShelleyBasedEraValidationError "Current treasury value" sbe)
+        (\cOnwards -> Right $ Just $ Featured cOnwards unTxCurrentTreasuryValue)
+        sbe
+
+validateTxTreasuryDonation :: ()
+  => ShelleyBasedEra era
+  -> Maybe TxTreasuryDonation
+  -> Either (TxNotSupportedInEraValidationError era) (Maybe (Featured ConwayEraOnwards era L.Coin))
+validateTxTreasuryDonation sbe mTreasuryDonation =
```

Comment:
same suggestion as above

Reply:
Since I rebased this PR on `main` (instead of https://github.com/IntersectMBO/cardano-cli/pull/778) this is irrelevant now, but I'll keep it mind when revisiting https://github.com/IntersectMBO/cardano-cli/pull/778 :+1: